### PR TITLE
[polars-sql] Adding SQL Context, SELECT and GROUP BY

### PR DESCRIPTION
**Pull request by potter420**
This is the another attempt at adding SQL support to `polars`, in separate crate.

This PR add:
- SQL Context Object, which currently support register DataFrame.
- SELECT, GROUP BY on a single DataFrame

### CODE REVIEW

**Comment by ritchie46** [comment]
Thanks a lot @potter420. Seems like a good start.

There are still a few comment outed lines that need to be removed. And it could use a `cargo fmt` and maybe a `cargo clippy` run.

File: polars-sql/src/context.rs [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+                        alias_map.insert(alias.name.value.clone(), tbl_name.to_owned());
+                    };
+                    tbl_name
+                } else {
+                    return Err(PolarsError::ComputeError(
+                        format!("Table name {tbl_name} was not found").into(),
+                    ));
+                }
+            }
+            // sqlparser::ast::TableFactor::Derived { lateral, subquery, alias } => todo!(),
```
Comment (ritchie46) [inline polars-sql/src/context.rs:49]: Can these comments be removed?

File: polars-sql/src/context.rs [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            Err(PolarsError::ComputeError(
+                "One and only one statement at a time please".into(),
+            ))
+        } else {
+            let ast = ast.get(0).unwrap();
+            Ok(match ast {
+                Statement::Query(query) => {
+                    let rs = match &query.body {
+                        SetExpr::Select(select_stmt) => self.execute_select(&*select_stmt)?,
+                        // SetExpr::Query(_) => todo!(), // Subqueries
```
Comment (ritchie46) [inline polars-sql/src/context.rs:177]: Can these be removed?

File: polars-sql/src/context.rs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+            final_proj_pos.sort_by(|(proj_pa, _), (proj_pb, _)| proj_pa.cmp(proj_pb));
+            // println!("after plan: {:?}", agg_df.schema(), );
+            let final_proj = final_proj_pos
+                .into_iter()
+                .map(|(_, shm_p)| col(agg_df.schema().get_index(shm_p).unwrap().0))
+                .collect::<Vec<_>>();
+            agg_df.select(final_proj)
+        };
+        // println!("after plan: {:?}", df.schema());
```
Comment (ritchie46) [inline polars-sql/src/context.rs:160]: Can these be removed?

File: polars-sql/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        "#,
+            )?
+            .collect()?;
+        let df_pl = df
+            .lazy()
+            .filter(col("a").gt(lit(10)).and(col("a").lt(lit(20))))
+            .select(&[col("a"), col("b"), (col("a") + col("b")).alias("c")])
+            .limit(100)
+            .collect()?;
+        println!("{:?}", df_sql);
```
Comment (ritchie46) [inline polars-sql/src/lib.rs:51]: Can these be removed?

**Review by ritchie46 — COMMENTED** [review]
Looks good! Can you remove the comments, then I merge it in.

**Comment by potter420** [comment]
Removed and ready for review

File: polars-sql/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  "ndarray",
+  "series_from_anyvalue",
+  "avro",
+  "parquet",
+  "ipc",
+  "is_in",
+  "serde",
+]
+
+[workspace]
```
Comment (messense) [inline polars-sql/Cargo.toml:72]: Any reason `polars-sql` defines its own workspace? It breaks building from sdist for `py-polars` now that `polars-sql` is included in it. See

* https://github.com/conda-forge/polars-feedstock/pull/50
* https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=573525&view=logs&jobId=6920d3d3-ef74-5584-b309-e8e970b1eb3b&j=6920d3d3-ef74-5584-b309-e8e970b1eb3b&t=bc2ce270-cdc3-50c0-3478-1273af1baa71

Comment (ritchie46) [reply polars-sql/Cargo.toml:72]: No particular reason. Good observation. I shall add that crate to the same workspace. :+1: 

So for an `sdist` package all local crates must share a workspace?

Comment (messense) [reply polars-sql/Cargo.toml:72]: > So for an sdist package all local crates must share a workspace?

At the moment I think so, it's a quirk from how maturin implements sdist with local path dependencies, I hope we can lift the limitation in the future.

**Pull request merged**